Support animated emojis and add google font's noto emojis

Turtlepaw opened the pull request:
### 🎥 Animated emojis
This pull request adds animated emojis (using lottie files):

![](https://raw.githubusercontent.com/Turtlepaw/flutter_emoji_feedback/main/msedge_d8jeGIRkrm.gif)

### New class for presets
> [!WARNING]
> Breaking change

This pull request also adds the new classes:

- [`StaticEmojiPreset`](https://github.com/Turtlepaw/flutter_emoji_feedback/blob/8a21ca896e01f11d254559c228176a48e2676e69/lib/src/models/preset.dart#L4)
- [`AnimatedEmojiPreset`](https://github.com/Turtlepaw/flutter_emoji_feedback/blob/8a21ca896e01f11d254559c228176a48e2676e69/lib/src/models/preset.dart#L11)

These both implement [`EmojiPreset`](https://github.com/Turtlepaw/flutter_emoji_feedback/blob/8a21ca896e01f11d254559c228176a48e2676e69/lib/src/models/base.dart#L3) and hold the emoji's of the preset along with idle emojis (specific to animated presets)

#### Small changes
- `EmojiFeedback` is now a stateful widget, removing the need for users to implement `rating` state
- The widget is now unfocused on start
- The example has been changed to show all presets and now has web support

#### Considerations
- ~~You may need to add an attribution notice for using Noto Emojis.~~ [`dcfcece`](https://github.com/stevenosse/flutter_emoji_feedback/pull/3/commits/dcfcece4b9cf516bd619c3d2b8540674df4d7e1e) attempts to solve this.
- This should be a major release so that pub doesn't automatically upgrade it
- The README should explain the breaking changes

Turtlepaw committed:
feat: add lottie animations and convert EmojiFeedback to a stateful widget

Turtlepaw committed:
feat: update README.md and widget is unselected by default

Turtlepaw committed:
feat: replace "good" with "neutral"

Turtlepaw committed:
feat(example): show all presets

Turtlepaw committed:
feat: add idle emojis and preset classes

Turtlepaw committed:
Include attribution to google for noto emojis

Turtlepaw committed:
Include attribution in LICENSE

Turtlepaw committed:
feat: add tap scaling

stevenosse commented on lib/src/emoji.dart:
```diff
@@ -0,0 +1,199 @@
+  @override
+  void didUpdateWidget(covariant EmojiItem oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    // if (widget.isActive && !oldWidget.isActive) {
```
If the commented code is no longer relevant, can it be removed?

stevenosse commented on lib/src/flutter_emoji_feedback.dart:
```diff
@@ -39,19 +40,13 @@ class EmojiFeedback extends StatelessWidget {
-  })  : assert(emojiPreset.length > 0),
+    this.onChangeWaitForAnimation = false,
+    this.tapScale = 0.6,
+  }) /* incompatible with new EmojiPreset class (I think)  : assert(minRating <= maxRating && maxRating <= emojiPreset.emojis.length),
```
Same, if something is not relevant anymore, i think it's better to get rid of it

stevenosse commented on lib/src/flutter_emoji_feedback.dart:
```diff
@@ -129,12 +124,31 @@ class EmojiFeedback extends StatelessWidget {
+  State<EmojiFeedback> createState() => _EmojiFeedbackStatefulState();
+}
+
+class _EmojiFeedbackStatefulState extends State<EmojiFeedback> {
```
```suggestion
class _EmojiFeedbackState extends State<EmojiFeedback> {
```

stevenosse commented on lib/src/emoji.dart:
```diff
@@ -0,0 +1,199 @@
+  }
+
+  Future<void> _initializeAnimation() async {
+    if (widget.emoji.src.endsWith(".json")) {
+      await Future.delayed(Duration.zero); // Wait for the widget to load
+      final composition =
+          await AssetLottie(widget.emoji.src, package: widget.emoji.package)
+              .load();
+      setState(() {
+        _animationDuration = composition.duration;
+        _controller.duration = composition.duration;
+      });
+    }
```
```suggestion
    WidgetsBinding.instance.addPostFrameCallback((_) {
        if (widget.emoji.src.endsWith(".json")) {
          final composition =
              await AssetLottie(widget.emoji.src, package: widget.emoji.package)
                  .load();
          setState(() {
            _animationDuration = composition.duration;
            _controller.duration = composition.duration;
          });
        }
```

stevenosse commented on lib/src/emoji.dart:
```diff
@@ -0,0 +1,199 @@
+      required this.onChangeWaitForAnimation,
+      required this.onSelected,
+      required this.idleEmoji,
+      required this.tapScale});
```
```suggestion
          required this.tapScale,
      });
```

stevenosse commented on .idx/dev.nix:
```diff
@@ -0,0 +1,68 @@
+# To learn more about how to use Nix to configure your environment
```
On the one hand, I'm wondering if we really want to track this file, but on the other, I'm thinking it might open the door to contributing to the project via IDX, and I think that's an interesting idea, what do you think?

Turtlepaw replied:
I think it's good to leave `.idx/*` files in the repository so that users, using Project IDX, don't have to create their own IDX config files and can quickly start contributing using an (already) working environment.

stevenosse reviewed (commented):
Thank you very much for this contribution, you have done a remarkable job!

I just have one more question. I haven't had time to run them yet, but are the existing tests still working?

Turtlepaw committed:
feat: add initialRating, fix tests, and resolve suggestions

Turtlepaw committed:
feat: add workflow

Turtlepaw committed:
feat: update README.md

Turtlepaw commented:
> Thank you very much for this contribution, you have done a remarkable job!
> 
> I just have one more question. I haven't had time to run them yet, but are the existing tests still working?

The tests now pass after fixing a couple of bugs, I've updated the README to document the recent changes, and I've also added a workflow to automatically run flutter tests.

[![](https://github.com/Turtlepaw/flutter_emoji_feedback/actions/workflows/test.yml/badge.svg)](https://github.com/Turtlepaw/flutter_emoji_feedback/actions/runs/12323697973)

Turtlepaw committed:
feat: improve README

Turtlepaw committed:
feat: add alt to github ci badge

Turtlepaw committed:
docs: Arrange emojis in a row in README

stevenosse commented:
LGTM 

If you are okay we can merge and release this new major version?

stevenosse commented:
LGTM

If you are okay we can merge and release this new major version?

Turtlepaw commented:
> If you are okay we can merge and release this new major version?

@stevenosse sounds good! 👍

Turtlepaw committed:
Improve README.md

stevenosse commented:
Release done: https://pub.dev/packages/flutter_emoji_feedback

Thank you very much for your contribution!
